core/src/oidc_auth.rs

```diff
@@ -40,9 +40,37 @@ pub enum Error {
     IssuerUriMustBeHttps,
     NoIdpServerInfo,
     CsrfMismatch,
+    MissingRefreshToken,
     Other(String),
 }
 
+macro_rules! handle_error {
+    () => {
+        |e| {
+            let msg = match e {
+                RequestTokenError::ServerResponse(provider_err) => {
+                    format!("Server returned error response: {:?}", provider_err)
+                }
+                RequestTokenError::Request(req) => {
+                    format!("Request failed: {:?}", req)
+                }
+                RequestTokenError::Parse(parse_err, res) => {
+                    let body = match std::str::from_utf8(&res) {
+                        Ok(text) => text.to_string(),
+                        Err(_) => format!("{:?}", &res),
+                    };
+                    format!(
+                        "Failed to parse server response: {} [response={:?}]",
+                        parse_err, body
+                    )
+                }
+                RequestTokenError::Other(msg) => msg,
+            };
+            Error::Other(format!("OpenID Connect: code exchange failed: {}", msg))
+        }
+    };
+}
+
 pub async fn do_auth_flow(params: CallbackContext) -> Result<IdpServerResponse, Error> {
     let idp_info = params.idp_info.ok_or(Error::NoIdpServerInfo)?;
```

Comment:
Creating a client is the same for both do_auth and do_refresh.
Can you move it out in a function which will return a client and the request_scopes.

Reply:
the refresh flow does not care about scopes